runtime/lua/vim/treesitter/query.lua

```diff
@@ -0,0 +1,176 @@
+local a = vim.api
+local lang = require'vim.treesitter.language'
+
+-- query: pattern matching on trees
+-- predicate matching is implemented in lua
+local Query = {}
+Query.__index = Query
+
+local M = {}
+
+local magic_prefixes = {['\\v']=true, ['\\m']=true, ['\\M']=true, ['\\V']=true}
+local function check_magic(str)
+  if string.len(str) < 2 or magic_prefixes[string.sub(str,1,2)] then
+    return str
+  end
+  return '\\v'..str
+end
+
+-- Some treesitter grammars extend others.
+-- We can use that to import the queries of the base language
+M.base_language_map = {
+  cpp = {'c'},
+  typescript = {'javascript'},
+  tsx = {'typescript', 'javascript'},
+}
+
+function M.parse_query(language, query)
+  lang.require_language(language)
+  local self = setmetatable({}, Query)
+  self.query = vim._ts_parse_query(language, vim.fn.escape(query,'\\'))
+  self.info = self.query:inspect()
+  self.captures = self.info.captures
+  self.regexes = {}
+  for id,preds in pairs(self.info.patterns) do
+    local regexes = {}
+    for i, pred in ipairs(preds) do
+      if (pred[1] == "match?" and type(pred[2]) == "number"
```

Comment:
Can you break this section into something like:

```lua
local predicate_handlers = {}
predicate_handlers["match?"] = function(predicate, other_required_params...)
  --   ... handling predicate stuff here
end
```
This way when we add more predicates it's obvious.

Reply:
Here is a thought that I had for a while : regexes are a pain in current implementation, they leverage vim builtin regexes while the whole treesitter runtime is in lua. I really think that using lua regexes is better, and can even provide better performances (here again, just a gut feeling, but the JIT can help here). Plus removing this and using lua regexes will simply remove this part of the code, and simplify the creation of queries (to go full lua), and have more flexibility.

Reply:
Would be nice if the predicate handler could automatically define the negation of the predicate. Like `match?` could automatically define `not-match?`

Reply:
Would be nice if the predicate handler could automatically define the negation of the predicate. Like `match?` could automatically define `not-match?`

Reply:
Plus Lua regexes can be multi-line

Reply:
Agree on the negation part for predicates. Regexes will be tackled later.